service/src/main/java/bio/terra/workspace/service/workspace/AwsCloudContextService.java

```diff
@@ -4,41 +4,46 @@
 import bio.terra.aws.resource.discovery.EnvironmentDiscovery;
 import bio.terra.aws.resource.discovery.Metadata;
 import bio.terra.workspace.app.configuration.external.AwsConfiguration;
-import bio.terra.workspace.app.configuration.external.FeatureConfiguration;
 import bio.terra.workspace.common.exception.InternalLogicException;
 import bio.terra.workspace.common.utils.AwsUtils;
 import bio.terra.workspace.db.WorkspaceDao;
+import bio.terra.workspace.service.features.FeatureService;
 import bio.terra.workspace.service.workspace.exceptions.CloudContextRequiredException;
 import bio.terra.workspace.service.workspace.exceptions.InvalidApplicationConfigException;
 import bio.terra.workspace.service.workspace.model.AwsCloudContext;
 import bio.terra.workspace.service.workspace.model.CloudPlatform;
+import com.google.common.base.Preconditions;
 import io.opencensus.contrib.spring.aop.Traced;
 import java.io.IOException;
 import java.util.*;
 import javax.validation.constraints.NotNull;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
-import org.springframework.util.Assert;
 
 /**
  * This service provides methods for managing AWS cloud context. These methods do not perform any
  * access control and operate directly against the {@link WorkspaceDao}
  */
 @Component
 public class AwsCloudContextService {
+  private final AwsConfiguration awsConfiguration;
   private final WorkspaceDao workspaceDao;
-  private final EnvironmentDiscovery environmentDiscovery;
+  private final FeatureService featureService;
+  private EnvironmentDiscovery environmentDiscovery;
 
   @Autowired
   public AwsCloudContextService(
-      WorkspaceDao workspaceDao,
-      FeatureConfiguration featureConfiguration,
-      AwsConfiguration awsConfiguration) {
+      WorkspaceDao workspaceDao, FeatureService featureService, AwsConfiguration awsConfiguration) {
+    this.awsConfiguration = awsConfiguration;
     this.workspaceDao = workspaceDao;
+    this.featureService = featureService;
     this.environmentDiscovery =
-        featureConfiguration.isAwsEnabled()
-            ? AwsUtils.createEnvironmentDiscovery(awsConfiguration)
-            : null;
+        featureService.awsEnabled() ? AwsUtils.createEnvironmentDiscovery(awsConfiguration) : null;
```

Comment:
i don't like how we are realizing on flag values at app start up. 

there might be a delay of flag fetching and app start up so we might be getting the default value here. also we could switch flag on independent of service redeployment now.

Reply:
We do want the `EnvironmentDiscovery` class to be long-lived, as it does some caching  to reduce cloud reads of infrequently-updated data in an S3 bucket.

So  perhaps the `maybeInitializeEnvironmentDiscovery()` method below should also be only call `AwsUtils. createEnvironmentDiscovery()` if `this.environmentDiscovery` is `null` AND AWS is enabled?

Reply:
done, dex suggested the same

btw i assume this is a known thing, i tried turn on aws and wsm doesn't build if it's on, complaning about discoveryConfig.

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'workspaceApiController' defined in file [/Users/yuhuyoyo/databiosphere/terra-workspace-manager/service/build/classes/java/main/bio/terra/workspace/app/controller/WorkspaceApiController.class]: Unsatisfied dependency expressed through constructor parameter 10; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'awsCloudContextService' defined in file [/Users/yuhuyoyo/databiosphere/terra-workspace-manager/service/build/classes/java/main/bio/terra/workspace/service/workspace/AwsCloudContextService.class]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [bio.terra.workspace.service.workspace.AwsCloudContextService]: Constructor threw exception; nested exception is java.lang.NullPointerException: Cannot invoke "bio.terra.workspace.app.configuration.external.AwsConfiguration$Discovery.getBucket()" because "discoveryConfig" is null
```

Reply:
We can only turn AWS on connected/integration tests when running in Verily `devel`, as that's the only environment with AWS backing.  Currently we only turn AWS on in `awsConnectedTests`, which do get run in downstream (`verily-src` repo) `pr-test.yaml` Github workflow.  

This workflow has a `workflow_dispatch` trigger, so you could trigger a run against a branch of the `verily-src` repo with the submodule pointed at your development branch.

Reply:
ah ok, i see you triggered it. I didn't turn the flag on in devel yet. i will flip it and re-run that test